software/contrib/cvecorder.py

```diff
@@ -0,0 +1,348 @@
+from europi import *
+from time import ticks_diff, ticks_ms
+from random import randint, uniform
+#from europi_script import EuroPiScript
+import machine
+import json
+import gc
+import micropython
+
+'''
+CVecorder
+author: Nik Ansell (github.com/gamecat69)
+date: 2022-04-04
+labels: sequencer, CV, performance
+
+Multi-channel CV recording and playback.
+
+Note, due to limitations in the RPi, only positive CV can be recorded and played back.
+
+Demo video: https://youtu.be/Crj0P7pr2YA
+
+digital_in: Clock input
+analog_in: Incoming CV
+
+button_1: Toggle recording stop/start
+button_2: Short Press: Cycle through CV recorder channels. Long Press (0.5 seconds): Clear the current bank. Long Press (2 seconds): Clear all banks.
+
+output_1: CV record / playback
+output_2: CV record / playback
+output_3: CV record / playback
+output_4: CV record / playback
+output_5: CV record / playback
+output_6: CV record / playback
+
+'''
+
+'''
+To do:
+- Add a morph capability using k2: left reduces CV values, right increases. Don’t adjust all values at once, so odds, then evens with each slight movement
+- Create documentation
+- Create upstream pull request
+'''
+
+# Needed if using europi_script
+#class CVecorder(EuroPiScript):
+class CVecorder():
+    def __init__(self):
+        
+        # Needed if using europi_script
+        super().__init__()
+
+        # Overclock the Pico for improved performance.
+        machine.freq(250_000_000)
+
+        # Initialize variables
+        self.step = 0
+        self.stepLength = 64
+        self.clockStep = 0
+        self.ActiveCvr = 0
+        self.ActiveBank = 0
+        self.resetTimeout = 500
+        self.debug = True
+        self.CvIn = 0
+        self.bankToSave = 0
+        self.debugTest = False
+
+        self.numCVR = 5  # Number of CV recorder channels - zero based
+        self.numCVRBanks = 5  # Number of CV recording channel banks - zero based
+
+        #self.CVR = []  # CV recorder channels
+        #self.CvRecording = []  # CV recorder flags
+
+        # Load CV Recordings from a previously stored state on disk or initialize if blank
+        self.loadState()
+
+        # Test routine, pick a random bank n times and save, then load the state
+        if self.debugTest:
+            print(micropython.mem_info("level"))
+            for n in range(3000):
+                # Clear vars
+                #self.CvRecording = []
+                print(f"Running test: {n}")
+                self.ActiveBank = randint(0, self.numCVRBanks)
+                self.ActiveCvr = randint(0, self.numCVR)
+                for i in range(0, self.stepLength-1):
+                    self.CVR[self.ActiveBank][self.ActiveCvr][i] = uniform(0.0, 9.99)
+                    #print(f"[{self.ActiveBank}][{self.ActiveCvr}][{i}] = {self.CVR[self.ActiveBank][self.ActiveCvr][i]}")
+                self.bankToSave = self.ActiveBank
+                self.saveState()
+                self.loadState()
```

Comment:
I would like to keep this code exactly where it is please, I do not want to spend more time creating a unit test at this point in time. If this will prevent my pull request so be it.